### PR TITLE
Add Deprecations

### PR DESCRIPTION
**Pull request by CSSFrancis**
#### Description of the change
Adds in Deprecation wrappers.

See Kikchipy, pyxem, hyperspy etc.

#### Progress of the PR
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [x] Unit tests with pytest for all lines
- [x] Clean code style by [running black](https://diffsims.readthedocs.io/en/latest/contributing.html#get-the-style-right)

#### Minimal example of the bug fix or new feature
```python
>>> from diffsims.utils._deprecated import deprecate
>>> @deprecate(since=0.8, removal=0.9, alternative="bar")
>>> def foo(self, n):
>>>     return n + 1

>>> @property
>>> @deprecate(since=0.9, removal=0.10, alternative="another", is_function=True)
>>> def this_property(self):
>>>     return 2
```

#### For reviewers

- [ ] The PR title is short, concise, and will make sense 1 year later.
- [ ] New functions are imported in corresponding `__init__.py`.
- [ ] New features, API changes, and deprecations are mentioned in the
      unreleased section in `CHANGELOG.rst`.
- [ ] Contributor(s) are listed correctly in `credits` in `diffsims/release_info.py` and
      in `.zenodo.json`.

### CODE REVIEW

**Comment by CSSFrancis** [comment]
@hakonanes Just adding these here so I can deprecate some code before tackling #198.

Should this be a Gist? or a small python package or??? I really like reusing this :)

**Comment by hakonanes** [comment]
> Should this be a Gist?

Perhaps. Would that make it easier to copy, somehow? Or do you mean in terms of visibility?

As mentioned before, I took the bits I wanted from Matplotlib's and scikit-image's deprecation tools. @harripj improved it after I added it to orix. Did you, @CSSFrancis, add functionality as well at some point? You have my blessing to put it up as a Gist if you want. But, you need @harripj's blessing as well. If you do make a Gist of the these deprecation tools, I'm OK with you licensing it in accordance with Matplotlib and scikit-image, that is not being restricted by GPLv3.

I don't think we should put a package on PyPI, as there are already at least deprecation packages on there:
* deprecation: https://pypi.org/project/deprecation/
* Deprecated: https://pypi.org/project/Deprecated/

File: CHANGELOG.rst
Comment (hakonanes) [inline CHANGELOG.rst]:
```suggestion
- Added deprecation tools for deprecating functions, parameters, methods, and properties.
```

File: CONTRIBUTING.rst
Comment (hakonanes) [inline CONTRIBUTING.rst]: These are code lines in a script rather than lines entered in a prompt, right? I suggest to remove the prompt.

```suggestion
    from diffsims.utils._deprecated import deprecate
    @deprecate(since=0.8, removal=0.9, alternative="bar")
    def foo(self, n):
        return n + 1

    @property
    @deprecate(since=0.9, removal=0.10, alternative="another", is_function=True)
    def this_property(self):
        return 2
```

File: diffsims/utils/_deprecated.py
Comment (hakonanes) [inline diffsims/utils/_deprecated.py]: Add blank lines for increased readability:

```suggestion
"""Helper functions and classes for managing diffsims.

This module and documentation is only relevant for diffsims developers,
not for users.

.. warning:
    This module and its submodules are for internal use only.  Do not
    use them in your own code. We may change the API at any time with no
    warning.
```

File: diffsims/utils/_deprecated.py
Comment (hakonanes) [inline diffsims/utils/_deprecated.py]:
```suggestion
        """Visible deprecation warning.
        
        Parameters
```

File: diffsims/utils/_deprecated.py
Comment (hakonanes) [inline diffsims/utils/_deprecated.py]:
```suggestion
    warning.

    Adapted from
```

File: diffsims/utils/_deprecated.py
Comment (hakonanes) [inline diffsims/utils/_deprecated.py]:
```suggestion
    signature.
    
    Adapted from `scikit-image
```

**Review by hakonanes — APPROVED** [review]
Looks good. I have a few minor suggestions that could be addressed before merging.

**Comment by harripj** [comment]
@hakonanes @CSSFrancis you have my blessing too. I wonder whether it is worth adding a reference back to orix somewhere in the Gist (header?)? I think that would be ample accreditation.

**Comment by hakonanes** [comment]
Can you make the suggested changes? I'll have a final look after that. Fairly certain we can merge after that again. I'd like to deprecate some old methods to get sets of {hkl} in #213.

**Comment by CSSFrancis** [comment]
pre-commit.ci autofix

**Comment by CSSFrancis** [comment]
Failing build is due to a Checksum not matching when using pip to install.  Kind of a new one :)

File: CHANGELOG.rst
Comment (hakonanes) [reply CHANGELOG.rst]: It appears you added a new line instead of replacing the current one?

File: CONTRIBUTING.rst
Comment (hakonanes) [inline CONTRIBUTING.rst]:
```suggestion
A deprecation decorator should be placed right above the object signature to be deprecated.

.. code-block:: python
    from diffsims.utils._deprecated import deprecated

    @deprecated(since=0.8, removal=0.9, alternative="bar")
    def foo(self, n):
        return n + 1

    @property
    @deprecated(since=0.9, removal=0.10, alternative="another", is_function=True)
```

**Review by hakonanes — CHANGES_REQUESTED** [review]

File: CONTRIBUTING.rst
Comment (CSSFrancis) [reply CONTRIBUTING.rst]: I guess I can competently use git some days...

**Review by hakonanes — APPROVED** [review]

**Pull request merged**
